clients/drcachesim/scheduler/scheduler.cpp

```diff
@@ -3288,32 +3493,58 @@ scheduler_tmpl_t<RecordType, ReaderType>::process_marker(input_info_t &input,
         input_ordinal_t target_idx = it->second;
         VPRINT(this, 3, "input %d re-scheduling input %d @%" PRIu64 "\n", input.index,
                target_idx, input.reader->get_last_timestamp());
-        // Release the input lock before acquiring sched_lock, to meet our lock
-        // ordering convention to avoid deadlocks.
+        // Release the input lock before acquiring more input locks
         input.lock->unlock();
         {
-            bool need_sched_lock;
-            auto scoped_sched_lock =
-                acquire_scoped_sched_lock_if_necessary(need_sched_lock);
             input_info_t *target = &inputs_[target_idx];
-            std::lock_guard<mutex_dbg_owned> lock(*target->lock);
-            if (target->unscheduled) {
+            std::unique_lock<mutex_dbg_owned> target_lock(*target->lock);
+            if (target->at_eof) {
+                VPRINT(this, 3, "input %d at eof ignoring re-schedule\n", target_idx);
+            } else if (target->unscheduled) {
                 target->unscheduled = false;
-                if (unscheduled_priority_.find(target)) {
-                    add_to_ready_queue(target);
-                    unscheduled_priority_.erase(target);
-                } else if (ready_priority_.find(target)) {
+                bool on_unsched_queue = false;
+                {
+                    std::lock_guard<mutex_dbg_owned> unsched_lock(
+                        *unscheduled_priority_.lock);
+                    if (unscheduled_priority_.queue.find(target)) {
+                        unscheduled_priority_.queue.erase(target);
+                        on_unsched_queue = true;
+                    }
+                }
+                // We have to give up the unsched lock before calling add_to_ready_queue
+                // as it acquires the output lock.
+                if (on_unsched_queue) {
+                    output_ordinal_t resume_output = target->prev_output;
+                    if (resume_output == INVALID_OUTPUT_ORDINAL)
+                        resume_output = output;
+                    // We can't hold any locks when calling add_to_ready_queue.
+                    target_lock.unlock();
+                    add_to_ready_queue(resume_output, target);
```

Comment:
Originally I had the input locks as "outer" to the output locks, meaning they were acquired first, but pop_ has to hold the output lock while looking at inputs so I reversed the order.  The downside is this operation and one other where an input lock has to be dropped.

I do feel the input locks in general might need improvement: more clearly specifying which fields they cover: there is a comment on that already.  We do run a race detector on analyzer code and it does report races on these data structures and locks (it found one in an earlier version of this code, and I've run it on large-ish inputs) so that gives some confidence.

I added here:

                    // This input is no longer on any queue, so few things can happen
                    // while we don't hold the input lock: a competing _SCHEDULE will
                    // not find the output and it can't have blocked_time>0 (invariant
                    // for things on unsched q); once it's on the new queue we don't
                    // do anything further here so we're good to go.